File: python/src/rag/vectorstore.py
import json
import math
from pathlib import Path
from typing import Any, Iterable, Union


def cosine_similarity(a: list[float], b: list[float]) -> float:
    if len(a) != len(b):
        raise ValueError("Embedding sizes do not match.")
    dot = sum(x * y for x, y in zip(a, b))
    norm_a = math.sqrt(sum(x * x for x in a))
    norm_b = math.sqrt(sum(y * y for y in b))
    if norm_a == 0 or norm_b == 0:
        return 0.0
    return dot / (norm_a * norm_b)


def load_store(path: Union[str, Path]) -> list[dict[str, Any]]:
    store_path = Path(path)
    if not store_path.exists():
        raise FileNotFoundError(f"Vector store not found: {store_path}")
    return json.loads(store_path.read_text(encoding="utf-8"))


def save_store(path: Union[str, Path], entries: Iterable[dict[str, Any]]) -> None:
    store_path = Path(path)
    store_path.parent.mkdir(parents=True, exist_ok=True)
    store_path.write_text(json.dumps(list(entries), indent=2), encoding="utf-8")


def search(
    entries: list[dict[str, Any]],
    query_embedding: list[float],
    top_k: int = 5,
) -> list[dict[str, Any]]:
    scored: list[dict[str, Any]] = []
    for entry in entries:
        score = cosine_similarity(query_embedding, entry["embedding"])
        scored.append({
            "id": entry["id"],
            "source": entry["source"],
            "text": entry["text"],
            "score": score,
        })

    scored.sort(key=lambda item: item["score"], reverse=True)
    return scored[:top_k]

File: node/src/rag/vectorstore.mjs
import fs from "fs/promises";
import path from "path";

// loadStore reads a JSON vector store from disk.
export async function loadStore(storePath) {
  const data = await fs.readFile(storePath, "utf-8");
  return JSON.parse(data);
}

// saveStore writes entries to a JSON vector store.
export async function saveStore(storePath, entries) {
  await fs.mkdir(path.dirname(storePath), { recursive: true });
  const payload = JSON.stringify(entries, null, 2);
  await fs.writeFile(storePath, payload, "utf-8");
}

// search returns the top-k most similar entries.
export function search(entries, queryEmbedding, topK = 5) {
  const scored = entries.map((entry) => ({
    id: entry.id,
    source: entry.source,
    text: entry.text,
    score: cosineSimilarity(queryEmbedding, entry.embedding),
  }));

  scored.sort((a, b) => b.score - a.score);
  return scored.slice(0, Math.max(0, topK));
}

function cosineSimilarity(a, b) {
  if (!Array.isArray(a) || !Array.isArray(b) || a.length !== b.length || a.length === 0) {
    return 0;
  }

  let dot = 0;
  let normA = 0;
  let normB = 0;
  for (let i = 0; i < a.length; i += 1) {
    dot += a[i] * b[i];
    normA += a[i] * a[i];
    normB += b[i] * b[i];
  }

  if (normA === 0 || normB === 0) {
    return 0;
  }
  return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}